### PR TITLE
Stabilize ptr::slice_from_raw_parts[_mut]

### LINKED ISSUE #36925: Tracking issue for `ptr::slice_from_raw_parts[_mut]`
In stable Rust, the only way to get a `*const [T]` with an arbitrary length is through `slice::from_raw_parts`. Unfortunately, that function returns a `&[T]`, which makes it very dubious to use with arbitrary raw pointers.

### PR DESCRIPTION
**Pull request by CAD97**
Closes #36925, the tracking issue.
Initial impl: #60667

r? @rust-lang/libs 

In addition to stabilizing, I've adjusted the example of `ptr::slice_from_raw_parts` to use `slice_from_raw_parts` instead of `slice_from_raw_parts_mut`, which was unnecessary for the example as written.

### CODE REVIEW

**Comment by CAD97** [comment]
Hm right let's inform the bots about the review

r? @oli-obk

**Comment by CAD97** [comment]
Actual error:

```
2020-01-14T21:53:12.3332167Z error: the feature `slice_from_raw_parts` has been stable since 1.42.0 and no longer requires an attribute to enable
2020-01-14T21:53:12.3333402Z --> src/liballoc/lib.rs:107:12
2020-01-14T21:53:12.3333973Z |
2020-01-14T21:53:12.3334715Z 107 | #![feature(slice_from_raw_parts)]
2020-01-14T21:53:12.3335426Z | ^^^^^^^^^^^^^^^^^^^^
2020-01-14T21:53:12.3335944Z |
2020-01-14T21:53:12.3336914Z = note: `-D stable-features` implied by `-D warnings`
2020-01-14T21:53:12.3337186Z
2020-01-14T21:53:12.3957867Z error: aborting due to previous error
2020-01-14T21:53:12.3958825Z
2020-01-14T21:53:12.4101399Z error: could not compile `alloc`.
2020-01-14T21:53:12.4102638Z
2020-01-14T21:53:12.4103349Z To learn more, run the command again with --verbose.
2020-01-14T21:53:12.4176865Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
2020-01-14T21:53:12.4177172Z Build completed unsuccessfully in 0:02:23
2020-01-14T21:53:12.4178280Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "-Zconfig-profile" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "2" "--release" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/src/libtest/Cargo.toml" "--message-format" "json-render-diagnostics"
2020-01-14T21:53:12.4178532Z expected success, got: exit code: 101
2020-01-14T21:53:12.4192562Z == clock drift check ==
2020-01-14T21:53:12.4201121Z local time: Tue Jan 14 21:53:12 UTC 2020
2020-01-14T21:53:12.6971069Z network time: Tue, 14 Jan 2020 21:53:12 GMT
2020-01-14T21:53:12.6975506Z == end clock drift check == 
```

cc @TimNN: error fragments did not include error message

(rebased)

**Comment by SimonSapin** [comment]
This is two free functions in the `ptr` module:

```rust
pub const fn slice_from_raw_parts<T>(data: *const T, len: usize) -> *const [T] {…}
pub const fn slice_from_raw_parts_mut<T>(data: *mut T, len: usize) -> *mut [T] {…}
```

@rfcbot fcp merge

**Comment by Amanieu** [comment]
@rfcbot concern &raw

I feel that the usefulness of this feature is tied with that of `&raw`: if in the end we decide to allow references to uninitialized data (as long as you don't deref them) then this function becomes much less useful since you can do the same thing with the existing `slice::from_raw_parts(_mut)` methods.

**Comment by CAD97** [comment]
@Amanieu the reason I want these methods actually has nothing to do with `&raw`; in fact, I know the allocation to be valid and initialized already and can use the reference functions freely, in theory. (I'm working with type erasing.)

The problem I'm dealing with is pointer provinence. I don't know (in my current generic API setup) whether the raw pointer I'm working with is from a `&T` or a `&mut T`.

If I use `slice::from_raw_parts`, then I'm manifesting a shared `&_` reference. Any pointer derived from this reference also has shared provenance, and cannot be used to write. This is obviously incorrect if the actual pointer I'm reconstructing should be able to write. You are never ever allowed to go `&_ -> &mut _`. This one is definitely wrong no matter the decision around `&raw` and references to trash.

If I use `slice::from_raw_parts_mut`, then I'm manifesting a unique `&mut _` reference. This one you _might_ be able to argue is potentially fine under a potential memory model, but that would require saying `&mut` isn't unique until it's dereferenced at a minimum.

And this is ignoring the case of potentially misaligned or null pointers, which are instant UB to use with the reference functions.

Ultimately, I think there are enough reasons to have these functions, even if they aren't strictly required by the memory model (for aligned non-null pointers), because they're a simple helper API when you're working with pointers to begin with that sidestep all of the thorny questions around references in pointer workloads that will remain no matter the specific memory model.

TL;DR: these functions are mainly useful when you don't know provenance (dependent on memory model), alignment, or null (definitely UB today).

**Comment by Amanieu** [comment]
I'm convinced!

@rfcbot resolve &raw

**Comment by CAD97** [comment]
I'm just going to go ahead and cc @rust-lang/wg-unsafe-code-guidelines just in case. This is an FCP-merge with some implication on unsafe guidelines.

**Comment by nikomatsakis** [comment]
For the record I think this stabilization makes sense. It doesn't seem to have many deep implications to me -- it is simply saying that one can construct a slice given a `*const T` and a `usize`. It's hard to imagine that not being true through various means (e.g., transmute etc), even if using them would otherwise be relying on unspecified implementation details. So yeah 👍 from me.

**Comment by RalfJung** [comment]
> I feel that the usefulness of this feature is tied with that of &raw: if in the end we decide to allow references to uninitialized data (as long as you don't deref them) then this function becomes much less useful since you can do the same thing with the existing slice::from_raw_parts(_mut) methods.

It still has some uses though -- you still need the raw methods for potentially dangling or out-of-bounds or NULL or unaligned pointers. That seems like a good enough motivation to me.

File: src/libcore/ptr/mod.rs [hunk 1]
```diff
@@ -248,17 +248,16 @@ pub(crate) struct FatPtr<T> {
...
 /// ```rust
-/// #![feature(slice_from_raw_parts)]
 /// use std::ptr;
 ///
 /// // create a slice pointer when starting out with a pointer to the first element
-/// let mut x = [5, 6, 7];
-/// let ptr = &mut x[0] as *mut _;
-/// let slice = ptr::slice_from_raw_parts_mut(ptr, 3);
+/// let x = [5, 6, 7];
+/// let ptr = &x[0] as *const _;
```
Comment (rkruppe) [inline src/libcore/ptr/mod.rs]: Currently miri considers this (more precisely, the access to an element other than the first in the line below) UB. See https://github.com/rust-lang/unsafe-code-guidelines/issues/134 for details and discussion whether it should be that way. Regardless of the outcome of that discussion, it seems better to side-step the issue by using `x.as_ptr()` here.

Comment (CAD97) [reply src/libcore/ptr/mod.rs]: Yes, `as_ptr` is definitely the correct choice here. I didn't even notice this error in the docs; good catch! (In my defense, I only removed the `mut` bit of what was already written.)

Comment (RalfJung) [reply src/libcore/ptr/mod.rs]: Good catch indeed! We really need to work on https://github.com/rust-lang/miri/issues/584...
